ksqldb-engine/src/main/java/io/confluent/ksql/schema/ksql/inference/DefaultSchemaInjector.java

```diff
@@ -176,6 +185,25 @@ private SchemaAndId getSchema(
     return result.schemaAndId.get();
   }
 
+  private static void validateSchemaInference(
+      final FormatInfo formatInfo,
+      final boolean isKey
+  ) {
+    /*
+     * Do validation when schemaId presents or we need to infer. Conditions to meet:
+     *  1. If schema id is provided, format must support schema inference
+     */
+
+    final String schemaIdName =
+        isKey ? CommonCreateConfigs.KEY_SCHEMA_ID : CommonCreateConfigs.VALUE_SCHEMA_ID;
+    if (!formatSupportsSchemaInference(formatInfo)) {
+      final String msg = (isKey ? CommonCreateConfigs.KEY_FORMAT_PROPERTY
+          : CommonCreateConfigs.VALUE_FORMAT_PROPERTY) + " should support schema inference "
+              + "when " + schemaIdName + " is provided!";
```

Comment:
This logic assumes that if we hit this error, then `SCHEMA_ID` was provided. What happens in the case where `SCHEMA_ID` is not provided but the format does not support schema inference? I'm guessing the code is set up in such a way that we will not reach this code path if that's the case (do we have tests for this?), but the logic in `getKeySchema()` and `getValueSchema()` above does not reflect this. This code is confusing as a result. Can we clean this up? 

Asked differently, what happens today when no key/value columns are provided and the key/value format doesn't support schema inference? Is `DefaultSchemaInjector` ever hit? If not, let's clean up the logic above to reflect that. If so, then this error message needs to be updated.

Reply:
There's this check before the validation in line 149:

```
    if (!props.getValueSchemaId().isPresent() && (hasValueElements(statement)
        || !formatSupportsSchemaInference(valueFormat))) {
```
So `DefaultSchemaInjector` will do nothing if no columns are provided and format doesn't support schema inference. https://github.com/confluentinc/ksql/blob/master/ksqldb-parser/src/main/java/io/confluent/ksql/parser/tree/TableElements.java#L81-L83 will hit when no key and value columns are provided somewhere later in the code. 

Let me check if no column and format doesn't support inference is tested and add test if not.

Reply:
I guess I can put the check in line 149 in this function and let this function return boolean indicating should do schema inference or not?

Reply:
The case is covered by:
https://github.com/confluentinc/ksql/blob/master/ksqldb-functional-tests/src/test/resources/query-validation-tests/elements.json#L1-L19
https://github.com/confluentinc/ksql/blob/master/ksqldb-engine/src/test/java/io/confluent/ksql/schema/ksql/inference/DefaultSchemaInjectorTest.java#L226-L248